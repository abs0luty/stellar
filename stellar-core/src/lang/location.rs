/// Represents location of an individual byte in a source file.
#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Location {
    line: u32,
    column: u32,
    index: u32,
}

impl Location {
    /// Creates a new location object.
    pub fn new(line: u32, column: u32, index: u32) -> Self {
        Self {
            line,
            column,
            index,
        }
    }

    /// Returns the location of the first byte in the file - Start Of File (SOF).
    pub fn sof() -> Self {
        Self {
            line: 1,
            column: 0,
            index: 0,
        }
    }

    pub fn line(&self) -> u32 {
        self.line
    }

    pub fn column(&self) -> u32 {
        self.column
    }

    pub fn index(&self) -> u32 {
        self.index
    }
}

/// Represents range of bytes in a source file.
#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Span {
    start: Location,
    end: Location,
}

impl Span {
    /// Creates a byte span.
    pub fn new(start: Location, end: Location) -> Self {
        Self { start, end }
    }

    /// Returns length of the span in bytes.
    pub fn len(&self) -> u32 {
        self.start.index - self.end.index
    }

    /// Returns location of the first byte in the span.
    pub fn start(&self) -> Location {
        self.start
    }

    /// Returns location of the last byte in the span.
    pub fn end(&self) -> Location {
        self.end
    }
}

/// Represents any object localized in a specific byte span 
/// of a source file.
pub trait Spanned {
    /// Returns a byte span associated with the object.
    fn span(&self) -> Span;
}
