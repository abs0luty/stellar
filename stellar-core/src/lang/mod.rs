mod cursor;

pub mod ast;
pub mod location;
pub mod parse;
pub mod scan;
pub mod token;
