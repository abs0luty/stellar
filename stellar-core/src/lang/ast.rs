use crate::lang::location::Span;

use super::{location::Spanned, token::Identifier};

#[derive(Debug, PartialEq)]
pub struct Property {
    pub name: Identifier,
    pub value: Expression
}

impl Spanned for Property {
    fn span(&self) -> Span {
        Span::new(self.name.span().start(), self.value.span().end())
    }
}

#[derive(Debug, PartialEq)]
pub enum Statement {
    Sequence { name: Identifier, block: Block },
    With { properties: Vec<Property>, block: Block },
    Expression(Expression)
}

#[derive(Debug, PartialEq)]
pub enum Expression {
    Float {
        value: f64,
        span: Span,
    },
    Integer {
        value: i64,
        span: Span,
    },
    Bool {
        value: bool,
        span: Span,
    },
    Binary {
        operator: BinaryOperator,
        left: Box<Expression>,
        right: Box<Expression>,
    },
    Prefix {
        operator: PrefixOperator,
        operand: Box<Expression>,
    },
}

impl Spanned for Expression {
    fn span(&self) -> Span {
        match self {
            Self::Binary { left, right, .. } => Span::new(left.span().start(), right.span().end()),
            Self::Prefix { operator, operand } => {
                Span::new(operator.span().start(), operand.span().end())
            }
            Self::Bool { span, .. } | Self::Float { span, .. } | Self::Integer { span, .. } => {
                *span
            }
        }
    }
}

#[derive(Debug, PartialEq)]
pub enum PrefixOperatorKind {
    Exclamation,
}

#[derive(Debug, PartialEq)]
pub struct PrefixOperator {
    pub kind: PrefixOperatorKind,
    pub span: Span,
}

impl Spanned for PrefixOperator {
    fn span(&self) -> Span {
        self.span
    }
}

#[derive(Debug, PartialEq)]
pub enum BinaryOperatorKind {
    Plus,
    Minus,
}

impl BinaryOperatorKind {
    pub fn precedence(&self) -> usize {
        match self {
            Self::Plus | Self::Minus => 1,
        }
    }
}

#[derive(Debug, PartialEq)]
pub struct BinaryOperator {
    pub kind: BinaryOperatorKind,
    pub span: Span,
}

impl BinaryOperator {
    pub fn precedence(&self) -> usize {
        self.kind.precedence()
    }
}

impl Spanned for BinaryOperator {
    fn span(&self) -> Span {
        self.span
    }
}

#[derive(Debug, PartialEq)]
pub struct Block {
    pub statements: Vec<Statement>,
    pub span: Span,
}
