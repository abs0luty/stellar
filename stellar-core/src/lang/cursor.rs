use std::{iter::Peekable, str::Chars};

use crate::lang::location::Location;

pub struct Cursor<'a> {
    source: &'a str,
    input: Peekable<Chars<'a>>,
    location: Location,
}

impl<'a> Cursor<'a> {
    pub fn new(source: &'a str) -> Self {
        Self {
            source,
            input: source.chars().peekable(),
            location: Location::sof(),
        }
    }

    pub fn source(&self) -> &'a str {
        return self.source;
    }

    pub fn location(&self) -> Location {
        return self.location;
    }

    pub fn peek(&mut self) -> Option<char> {
        return self.input.peek().copied();
    }

    pub fn next(&mut self) -> Option<char> {
        if let Some(&next_char) = self.input.peek() {
            self.location = if next_char == '\n' {
                Location::new(
                    self.location.line() + 1,
                    0,
                    self.location.index() + next_char.len_utf8() as u32,
                )
            } else {
                Location::new(
                    self.location.line(),
                    self.location.column() + 1,
                    self.location.index() + next_char.len_utf8() as u32,
                )
            };
        }

        return self.input.next();
    }
}
