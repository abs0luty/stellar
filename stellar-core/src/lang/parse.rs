use crate::lang::{
    ast::{Block, Expression, Statement},
    location::{Span, Spanned},
    token::{Keyword, Punctuation, Token, TokenStream, TokenStreamCursor},
};

use super::{ast::Property, token::Identifier};

/// Parses a given token stream into Abstract Syntax Tree (AST).
pub fn parse(stream: TokenStream) -> Result<Vec<Statement>, ParseError> {
    let Some(mut cursor) = stream.into_cursor() else {
        return Err(ParseError::InvalidTokenStream);
    };
    let mut statements = Vec::new();

    loop {
        skip_eols(&mut cursor);

        if cursor.peek().is_eof() {
            break;
        }

        statements.push(parse_statement(&mut cursor)?);
    }

    Ok(statements)
}

fn parse_block(cursor: &mut TokenStreamCursor) -> Result<Block, ParseError> {
    let start = parse_punctuation(cursor, Punctuation::LeftBrace)?
        .span()
        .start(); // '{'
    let mut statements = Vec::new();

    loop {
        skip_eols(cursor);

        if cursor.peek().is_punctuation(Punctuation::RightBrace) {
            break;
        }

        statements.push(parse_statement(cursor)?);
    }

    let end = cursor.next().span().end(); // '}'

    Ok(Block {
        statements,
        span: Span::new(start, end),
    })
}

fn parse_statement(cursor: &mut TokenStreamCursor) -> Result<Statement, ParseError> {
    match cursor.peek() {
        token if token.is_keyword(Keyword::Sequence) => parse_sequence_statement(cursor),
        token if token.is_keyword(Keyword::With) => parse_with_statement(cursor),
        _ => parse_expression(cursor, 0).map(|e| Statement::Expression(e)),
    }
}

fn parse_property(cursor: &mut TokenStreamCursor) -> Result<Property, ParseError> {
    let name = parse_identifier(cursor)?; // name
    parse_punctuation(cursor, Punctuation::Colon)?; // ':'

    let value = parse_expression(cursor, 0)?;

    Ok(Property { name, value })
}

fn parse_with_statement(cursor: &mut TokenStreamCursor) -> Result<Statement, ParseError> {
    cursor.next(); // 'with'

    let mut properties = Vec::new();

    properties.push(parse_property(cursor)?);

    while cursor.peek().is_punctuation(Punctuation::Comma) {
        cursor.next();

        if cursor.peek().is_punctuation(Punctuation::LeftBrace) {
            break; // with a: test, b: 3, { } - still counts
        }

        properties.push(parse_property(cursor)?)
    }

    let block = parse_block(cursor)?;

    Ok(Statement::With { properties, block })
}

fn parse_expression(
    cursor: &mut TokenStreamCursor,
    precedence: usize,
) -> Result<Expression, ParseError> {
    let mut left = parse_prefix_expression(cursor)?;

    while let Some(operator) = cursor.peek().into_binary_operator() {
        let operator_precedence = operator.precedence();

        if operator_precedence < precedence {
            break;
        }

        cursor.next();

        // 1 |  3 + 
        // 2 |  2 # Expression is continued on the new line  
        skip_eols(cursor);

        let right = parse_expression(cursor, operator_precedence + 1)?;
        left = Expression::Binary {
            left: Box::new(left),
            operator,
            right: Box::new(right),
        };
    }

    Ok(left)
}

fn parse_prefix_expression(cursor: &mut TokenStreamCursor) -> Result<Expression, ParseError> {
    match cursor.peek() {
        Token::Integer { value, span } => {
            cursor.next();

            Ok(Expression::Integer { value, span })
        }
        Token::Float { value, span } => {
            cursor.next();

            Ok(Expression::Float { value, span })
        }
        token if token.is_punctuation(Punctuation::LeftParen) => {
            cursor.next(); // '('

            let expression = parse_expression(cursor, 0)?;

            parse_punctuation(cursor, Punctuation::RightParen)?; // ')'

            Ok(expression)
        }
        token => {
            let Some(prefix_operator) = token.into_prefix_operator() else {
                return Err(ParseError::ExpectedExpression { token });
            };

            let operand = parse_expression(cursor, usize::MAX)?;

            Ok(Expression::Prefix {
                operator: prefix_operator,
                operand: Box::new(operand),
            })
        }
    }
}

fn parse_sequence_statement(cursor: &mut TokenStreamCursor) -> Result<Statement, ParseError> {
    cursor.next(); // 'sequence' keyword

    let name = parse_identifier(cursor)?;
    let block = parse_block(cursor)?;

    Ok(Statement::Sequence { name, block })
}

fn parse_identifier(cursor: &mut TokenStreamCursor) -> Result<Identifier, ParseError> {
    let got = cursor.next();
    let Token::Identifier(identifier) = got else {
        return Err(ParseError::ExpectedIdentifier { got });
    };

    Ok(identifier)
}

fn parse_punctuation(
    cursor: &mut TokenStreamCursor,
    punctuation: Punctuation,
) -> Result<Token, ParseError> {
    let got = cursor.next();
    if !got.is_punctuation(punctuation) {
        return Err(ParseError::ExpectedPunctuation {
            expected: punctuation,
            got,
        });
    }

    Ok(got)
}

fn skip_eols(cursor: &mut TokenStreamCursor) {
    while cursor.peek().is_eol() {
        cursor.next();
    }
}

#[derive(Debug)]
pub enum ParseError {
    InvalidTokenStream,
    ExpectedExpression { token: Token },
    ExpectedIdentifier { got: Token },
    ExpectedPunctuation { expected: Punctuation, got: Token },
}

#[cfg(test)]
mod tests {
    use lasso::Rodeo;

    use crate::lang::{
        ast::{Block, Expression, Property, Statement},
        location::{Location, Span},
        scan::scan,
        token::Identifier,
    };

    use super::parse;

    #[test]
    fn test_empty() {
        let mut rodeo = Rodeo::new();
        let token_stream = scan("", &mut rodeo).unwrap();
        let statements = parse(token_stream).unwrap();

        assert!(statements.is_empty());
    }

    #[test]
    fn test_with() {
        let mut rodeo = Rodeo::new();
        let token_stream = scan("with a: 3, b: 4, {}", &mut rodeo).unwrap();
        let statements = parse(token_stream).unwrap();

        assert_eq!(
            statements.first().unwrap(),
            &Statement::With {
                properties: vec![
                    Property {
                        name: Identifier::new(
                            rodeo.get_or_intern("a"),
                            Span::new(Location::new(1, 5, 5), Location::new(1, 6, 6))
                        ),
                        value: Expression::Integer {
                            value: 3,
                            span: Span::new(Location::new(1, 8, 8), Location::new(1, 9, 9))
                        }
                    },
                    Property {
                        name: Identifier::new(
                            rodeo.get_or_intern("b"),
                            Span::new(Location::new(1, 11, 11), Location::new(1, 12, 12))
                        ),
                        value: Expression::Integer {
                            value: 4,
                            span: Span::new(Location::new(1, 14, 14), Location::new(1, 15, 15))
                        }
                    }
                ],
                block: Block {
                    statements: vec![],
                    span: Span::new(Location::new(1, 17, 17), Location::new(1, 19, 19))
                }
            }
        );
    }

    #[test]
    fn test_sequence() {
        let mut rodeo = Rodeo::new();
        let token_stream = scan("sequence name {}", &mut rodeo).unwrap();
        let statements = parse(token_stream).unwrap();

        assert_eq!(
            statements.first().unwrap(),
            &Statement::Sequence {
                name: Identifier::new(
                    rodeo.get_or_intern("name"),
                    Span::new(Location::new(1, 9, 9), Location::new(1, 13, 13))
                ),
                block: Block {
                    statements: Vec::new(),
                    span: Span::new(Location::new(1, 14, 14), Location::new(1, 16, 16))
                }
            }
        );
    }
}
