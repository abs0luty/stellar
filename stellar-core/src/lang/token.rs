use lasso::Spur;

use crate::lang::location::{Location, Span, Spanned};

use super::ast::{BinaryOperator, BinaryOperatorKind, PrefixOperator, PrefixOperatorKind};

#[derive(Debug, PartialEq, Copy, Clone)]
pub enum Keyword {
    With,
    Wait,
    Sequence,
}

#[derive(Debug, PartialEq, Copy, Clone)]
pub enum Punctuation {
    Exclamation,
    LeftBrace,
    RightBrace,
    LeftBracket,
    RightBracket,
    LeftParen,
    RightParen,
    Plus,
    Minus,
    PlusEq,
    MinusEq,
    Dot,
    Colon,
    Comma,
}

impl Punctuation {
    pub fn into_binary_operator_kind(&self) -> Option<BinaryOperatorKind> {
        match self {
            Self::Plus => Some(BinaryOperatorKind::Plus),
            Self::Minus => Some(BinaryOperatorKind::Minus),
            _ => None,
        }
    }

    pub fn into_prefix_operator_kind(&self) -> Option<PrefixOperatorKind> {
        match self {
            Self::Exclamation => Some(PrefixOperatorKind::Exclamation),
            _ => None,
        }
    }
}

#[derive(Debug, PartialEq, Copy, Clone)]
pub struct Identifier {
    name: Spur,
    span: Span,
}

impl Identifier {
    pub fn new(name: Spur, span: Span) -> Self {
        Self { name, span }
    }

    pub fn name(&self) -> Spur {
        self.name
    }
}

impl Spanned for Identifier {
    fn span(&self) -> Span {
        self.span
    }
}

#[derive(Debug, PartialEq, Copy, Clone)]
pub enum Token {
    Keyword {
        keyword: Keyword,
        span: Span,
    },
    Identifier(Identifier),
    Punctuation {
        punctuation: Punctuation,
        span: Span,
    },
    Float {
        value: f64,
        span: Span,
    },
    Integer {
        value: i64,
        span: Span,
    },
    Bool {
        value: bool,
        span: Span,
    },
    EOL {
        span: Span,
    },
    EOF {
        /// Location of the last byte in the source file.
        location: Location,
    },
}

impl Token {
    pub fn is_eof(&self) -> bool {
        matches!(self, Self::EOF { .. })
    }

    pub fn is_keyword(&self, keyword: Keyword) -> bool {
        match self {
            Self::Keyword {
                keyword: my_keyword,
                ..
            } => keyword == *my_keyword,
            _ => false,
        }
    }

    pub fn is_punctuation(&self, punctuation: Punctuation) -> bool {
        match self {
            Self::Punctuation {
                punctuation: my_punctuation,
                ..
            } => punctuation == *my_punctuation,
            _ => false,
        }
    }

    pub fn into_binary_operator(&self) -> Option<BinaryOperator> {
        match self {
            Self::Punctuation { punctuation, span } => {
                let kind = punctuation.into_binary_operator_kind()?;

                Some(BinaryOperator { kind, span: *span })
            }
            _ => None,
        }
    }

    pub fn into_prefix_operator(&self) -> Option<PrefixOperator> {
        match self {
            Self::Punctuation { punctuation, span } => {
                let kind = punctuation.into_prefix_operator_kind()?;

                Some(PrefixOperator { kind, span: *span })
            }
            _ => None,
        }
    }

    pub fn is_identifier(&self) -> bool {
        matches!(self, Self::Identifier { .. })
    }
}

impl Spanned for Token {
    fn span(&self) -> Span {
        match self {
            Self::EOF { location } => Span::new(
                *location,
                Location::new(location.line(), location.column() + 1, location.index() + 1),
            ),
            Self::Identifier(Identifier { span, .. })
            | Self::Punctuation { span, .. }
            | Self::Keyword { span, .. }
            | Self::Integer { span, .. }
            | Self::Float { span, .. }
            | Self::Bool { span, .. }
            | Self::EOL { span } => *span,
        }
    }
}

#[derive(Debug, PartialEq)]
pub struct TokenStream(Vec<Token>);

impl TokenStream {
    /// Creates a new empty token stream.
    pub fn new() -> Self {
        Self(Vec::new())
    }

    /// Appends a token to the stream.
    pub fn push(&mut self, token: Token) {
        self.0.push(token);
    }

    /// Returns token with a specified index in the stream. In case index
    /// is out of bounds, EOF token (End Of File) is returned.
    fn get(&self, index: usize) -> Token {
        if index > self.0.len() {
            self.0.last().copied().unwrap_or(Token::EOF {
                location: Location::sof(),
            })
        } else {
            self.0[index]
        }
    }

    /// Returns a cursor over the token stream. See [`TokenStreamCursor`] for more details.
    pub fn into_cursor(self) -> Option<TokenStreamCursor> {
        // Ensure last token is EOF.
        if self.0.last().map_or(true, |maybe_eof| !maybe_eof.is_eof()) {
            return None;
        }

        Some(TokenStreamCursor::new(self))
    }
}

/// A cursor for navigating through a stream of tokens.
///
/// This struct provides functionality to sequentially traverse
/// a [`TokenStream`], allowing you to retrieve tokens one at a time
/// or peek at the upcoming token without advancing the cursor.
/// It tracks the current position in the stream and ensures
/// that an **EOF (End Of File) token is returned when no more
/// tokens are available**.
pub struct TokenStreamCursor {
    stream: TokenStream,
    location: usize,
}

impl TokenStreamCursor {
    fn new(stream: TokenStream) -> Self {
        Self {
            stream,
            location: 0,
        }
    }

    /// Retrieves the next token from the stream, advancing the cursor
    /// to the subsequent position. If no more tokens are available,
    /// an EOF (End Of File) token is returned.
    pub fn next(&mut self) -> Token {
        self.location += 1;

        self.stream.get(self.location - 1)
    }

    /// Provides a glimpse of the next token without advancing the cursor
    /// (compared to [`TokenStreamCursor::next`]). If no more tokens are 
    /// available, an EOF (End Of File) token is returned.
    pub fn peek(&mut self) -> Token {
        self.stream.get(self.location)
    }
}

impl IntoIterator for TokenStream {
    type IntoIter = std::vec::IntoIter<Self::Item>;
    type Item = Token;

    fn into_iter(self) -> Self::IntoIter {
        self.0.into_iter()
    }
}