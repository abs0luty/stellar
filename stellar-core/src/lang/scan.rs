use lasso::Rodeo;

use crate::lang::{
    cursor::Cursor,
    location::{Span, Spanned},
    token::{Keyword, Punctuation, Token, TokenStream},
};

use super::token::Identifier;

/// Scans a given Stellar source text and converts into a vector of tokens.
///
/// # Errors
/// In case any obvious syntax errors, which affected the scanning process
/// were found, [`ScannerError`] will be returned.
pub fn scan(source: &str, rodeo: &mut Rodeo) -> Result<TokenStream, ScanError> {
    let mut cursor = Cursor::new(source);
    let mut stream = TokenStream::new();

    loop {
        let token = scan_next_token(&mut cursor, rodeo)?;

        if let Token::EOF { .. } = token {
            stream.push(token);

            break;
        }

        stream.push(token);
    }

    Ok(stream)
}

macro_rules! match_punctuation {
    ($char:expr, $cursor:expr, $start:expr,
        { $($single:pat => $single_token:expr,)+ },
        { $($first:pat,$second:pat => $pair_token:expr,)+ }) => {{
        match ($char, $cursor.peek()) {
            // Handle two-character punctuations
            $(
                ($first, Some($second)) => {
                    $cursor.next();

                    Ok(Token::Punctuation {
                        punctuation: $pair_token,
                        span: Span::new($start, $cursor.location()),
                    })
                },
            )+
            // Handle single-character punctuations
            $(
                ($single, _) => Ok(Token::Punctuation {
                    punctuation: $single_token,
                    span: Span::new($start, $cursor.location()),
                }),
            )+
            // Handle unexpected character
            _ => Err(ScanError::UnexpectedCharacter {
                character: $char,
                span: Span::new($start, $cursor.location()),
            }),
        }
    }};
}

/// Scans the next token in the source text and advances position of the [`Cursor`].
fn scan_next_token(cursor: &mut Cursor, rodeo: &mut Rodeo) -> Result<Token, ScanError> {
    while let Some(c) = cursor.peek() {
        match c {
            // Skip whitespace (except line breaks).
            c if c.is_whitespace() && c != '\n' => {
                cursor.next();
            }
            // Skip single-line comments starting with '#'.
            '#' => {
                while let Some(c) = cursor.next() {
                    if c == '\n' {
                        break; // Stop skipping at the end of the line.
                    }
                }
            }
            _ => break, // Stop skipping when non-whitespace and non-comment are found.
        }
    }

    let Some(c) = cursor.peek() else {
        return Ok(Token::EOF {
            location: cursor.location(),
        });
    };

    match c {
        '\n' => {
            let location = cursor.location();
            cursor.next();

            Ok(Token::EOL {
                span: Span::new(location, cursor.location()),
            })
        }
        c if c.is_alphabetic() || c == '_' => Ok(scan_name(cursor, rodeo)),
        c if c.is_numeric() || c == '.' => Ok(scan_number_or_dot(cursor)),
        _ => {
            let start = cursor.location();
            cursor.next();

            match_punctuation!(
                c,
                cursor,
                start,
                {
                    '-' => Punctuation::Minus,
                    '+' => Punctuation::Plus,
                    '*' => Punctuation::Star,
                    '/' => Punctuation::Slash,
                    '{' => Punctuation::LeftBrace,
                    '}' => Punctuation::RightBrace,
                    '[' => Punctuation::LeftBracket,
                    ']' => Punctuation::RightBracket,
                    '(' => Punctuation::LeftParen,
                    ')' => Punctuation::RightParen,
                    ':' => Punctuation::Colon,
                    '.' => Punctuation::Dot,
                    ',' => Punctuation::Comma,
                },
                {
                    '-', '=' => Punctuation::MinusEq,
                    '+', '=' => Punctuation::PlusEq,
                }
            )
        }
    }
}

/// Scans the next candidate for identifier token in the source text and if
/// its name matches any known keywords returns keyword token.
fn scan_name(cursor: &mut Cursor, rodeo: &mut Rodeo) -> Token {
    let mut name = String::new();
    let start = cursor.location();

    while let Some(c) = cursor.peek() {
        if !c.is_alphanumeric() || c == '_' {
            break;
        }

        name.push(c);
        cursor.next();
    }

    let span = Span::new(start, cursor.location());

    match name.as_str() {
        "with" => Token::Keyword {
            keyword: Keyword::With,
            span,
        },
        "wait" => Token::Keyword {
            keyword: Keyword::Wait,
            span,
        },
        "sequence" => Token::Keyword {
            keyword: Keyword::Sequence,
            span,
        },
        "play" => Token::Keyword {
            keyword: Keyword::Play,
            span,
        },
        "true" => Token::Bool { value: true, span },
        "false" => Token::Bool { value: false, span },
        _ => Token::Identifier(Identifier::new(rodeo.get_or_intern(name), span)),
    }
}

/// Scans a number or a dot (`.`) from the source text.
fn scan_number_or_dot(cursor: &mut Cursor) -> Token {
    let start = cursor.location();
    let mut has_dot = false;

    while let Some(c) = cursor.peek() {
        if c.is_numeric() {
            cursor.next();
        } else if c == '.' && !has_dot {
            has_dot = true;
            cursor.next();
        } else {
            break;
        }
    }

    let end = cursor.location();

    if end.index() - start.index() == 1 && has_dot {
        return Token::Punctuation {
            punctuation: Punctuation::Dot,
            span: Span::new(start, end),
        };
    }

    let lexeme = &cursor.source()[(start.index() as usize)..(end.index() as usize)];

    if has_dot {
        Token::Float {
            value: lexeme.parse::<f64>().unwrap(),
            span: Span::new(start, end),
        }
    } else {
        Token::Integer {
            value: lexeme.parse::<i64>().unwrap(),
            span: Span::new(start, end),
        }
    }
}

#[derive(Debug, PartialEq)]
pub enum ScanError {
    UnexpectedCharacter { character: char, span: Span },
}

impl Spanned for ScanError {
    fn span(&self) -> Span {
        match self {
            Self::UnexpectedCharacter { span, .. } => *span,
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use lasso::Rodeo;

    use super::scan;

    macro_rules! test_scan {
        ($(($name:ident, $source:expr)),* $(,)?) => {
            $(
                #[test]
                fn $name() {
                    let mut rodeo = Rodeo::new();
                    assert_debug_snapshot!(scan($source, &mut rodeo));
                }
            )*
        };
    }

    test_scan!(
        (eof, ""),
        (unexpected_char, "!"),
        (punctuation, "("),
        (number_and_dot, "3 3.2."),
        (name, "wait time"),
    );
}
