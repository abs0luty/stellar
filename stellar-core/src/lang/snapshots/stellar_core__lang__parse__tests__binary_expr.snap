---
source: stellar-core/src/lang/parse.rs
expression: parse(token_stream)
snapshot_kind: text
---
Ok(
    [
        Expression(
            Binary {
                operator: BinaryOperator {
                    kind: Minus,
                    span: Span {
                        start: Location {
                            line: 2,
                            column: 13,
                            index: 18,
                        },
                        end: Location {
                            line: 2,
                            column: 14,
                            index: 19,
                        },
                    },
                },
                left: Binary {
                    operator: BinaryOperator {
                        kind: Plus,
                        span: Span {
                            start: Location {
                                line: 1,
                                column: 2,
                                index: 2,
                            },
                            end: Location {
                                line: 1,
                                column: 3,
                                index: 3,
                            },
                        },
                    },
                    left: Identifier(
                        Identifier {
                            name: Spur(1),
                            span: Span {
                                start: Location {
                                    line: 1,
                                    column: 0,
                                    index: 0,
                                },
                                end: Location {
                                    line: 1,
                                    column: 1,
                                    index: 1,
                                },
                            },
                        },
                    ),
                    right: Binary {
                        operator: BinaryOperator {
                            kind: Star,
                            span: Span {
                                start: Location {
                                    line: 2,
                                    column: 3,
                                    index: 8,
                                },
                                end: Location {
                                    line: 2,
                                    column: 4,
                                    index: 9,
                                },
                            },
                        },
                        left: Integer {
                            value: 2,
                            span: Span {
                                start: Location {
                                    line: 2,
                                    column: 1,
                                    index: 6,
                                },
                                end: Location {
                                    line: 2,
                                    column: 2,
                                    index: 7,
                                },
                            },
                        },
                        right: Binary {
                            operator: BinaryOperator {
                                kind: Plus,
                                span: Span {
                                    start: Location {
                                        line: 2,
                                        column: 8,
                                        index: 13,
                                    },
                                    end: Location {
                                        line: 2,
                                        column: 9,
                                        index: 14,
                                    },
                                },
                            },
                            left: Integer {
                                value: 3,
                                span: Span {
                                    start: Location {
                                        line: 2,
                                        column: 6,
                                        index: 11,
                                    },
                                    end: Location {
                                        line: 2,
                                        column: 7,
                                        index: 12,
                                    },
                                },
                            },
                            right: Identifier(
                                Identifier {
                                    name: Spur(2),
                                    span: Span {
                                        start: Location {
                                            line: 2,
                                            column: 10,
                                            index: 15,
                                        },
                                        end: Location {
                                            line: 2,
                                            column: 11,
                                            index: 16,
                                        },
                                    },
                                },
                            ),
                        },
                    },
                },
                right: Integer {
                    value: 3,
                    span: Span {
                        start: Location {
                            line: 2,
                            column: 15,
                            index: 20,
                        },
                        end: Location {
                            line: 2,
                            column: 16,
                            index: 21,
                        },
                    },
                },
            },
        ),
    ],
)
